azurelinuxagent/common/protocol/wire.py

```diff
@@ -998,6 +983,62 @@ def check_wire_protocol_version(self):
                      "advised by Fabric.").format(PROTOCOL_VERSION)
             raise ProtocolNotFoundError(error)
 
+    def send_request_using_appropriate_channel(self, direct_func, host_func):
+        # A wrapper method for all function calls that send HTTP requests. The purpose of the method is to
+        # define which channel to use, direct or through the host plugin. For the host plugin channel,
+        # also implement a retry mechanism.
+
+        # By default, the direct channel is the default channel. If that is the case, try getting a response
+        # through that channel. On failure, fall back to the host plugin channel.
+
+        # When using the host plugin channel, regardless if it's set as default or not, try sending the request first.
+        # On specific failures that indicate a stale goal state (such as resource gone or invalid container parameter),
+        # refresh the goal state and try again. If successful, set the host plugin channel as default. If failed,
+        # raise the exception.
+
+        if not HostPluginProtocol.is_default_channel():
```

Comment:
Initially we used to retry both direct download and via host on a ResourceGoneError,  now we just do it if we get the error via Host. 
Could there be a scenario where we'd need to retry on RGE during direct download?

Reply:
We actually didn't retry on `ResourceGoneError` via default channel. If you take a look at the previous implementation of the function `get_artifacts_profile`, you'll see that on `ResourceGoneError` we set the host channel as default and go in the loop again, which will force a goal state refresh and then a fetch using the host plugin.